importer/importer.go

```diff
@@ -0,0 +1,182 @@
+	i := &KindaFastImporter{
+		fastImporter: importer.Default().(types.ImporterFrom),
+		cwd: cwd,
+		packages: make(map[pkgIdent]*PackageInfo),
+		fileSet: token.NewFileSet(),
+		log: log,
+	}
+	// config references the importer, so initialize it after
+	i.config = &types.Config{
+		FakeImportC:      true,
+		Importer:         i,
+	}
+
+	return i, i
+}
+
+func (i *KindaFastImporter) Import(path string) (*types.Package, error) {
+	return i.ImportFrom(path, "", 0)
+}
+
+func (i *KindaFastImporter) ImportFrom(path, srcDir string, mode types.ImportMode) (*types.Package, error) {
+	log := i.log.WithValues("import path", path, "source directory", srcDir)
+	log.V(1).Info("trying to import package")
+
+	// first, double-check that we haven't already loaded this
+	ident := pkgIdent{path: path}
+	if pkg, present := i.packages[ident]; present {
+		if pkg == nil {
+			return nil, fmt.Errorf("import loop detected for package %s", ident)
+		}
+		log.V(1).Info("package already imported")
+		return pkg.ParseInfo, nil
+	}
+
+	// mark that we're checking to avoid loops
+	i.packages[ident] = nil
+
+	// next, try fast import:
+	// figure out where srcDir actually is...
+	fullDir := pathpkg.Join(i.cwd, srcDir)
+	// ...and fast-import from it
+	pkg, err := i.fastImporter.ImportFrom(path, fullDir, mode)
+	if err == nil {
+		// yay, fast import!
+		log.V(1).Info("fast-imported package")
+		buildPkg, err := build.Default.Import(path, fullDir, build.AllowBinary | build.FindOnly)
+		if err != nil {
+			return nil, err
+		}
+		i.packages[ident] = &PackageInfo{
+			BuildInfo: buildPkg,
+			ParseInfo: pkg,
+		}
+		return pkg, nil
+	}
+	
+	// otherwise, import the package from source
+	log.V(1).Info("unable to fast-import package, importing from source", "error", err)
+
+	// import, grabbing the build information
+	buildPkg, err := build.Default.Import(path, fullDir, build.AllowBinary)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO(directxman12): support xtest and test files
+	// parse the files in the package...
+	files, err := i.parsePackageFiles(buildPkg)
+	if err != nil {
+		return nil, err
+	}
+	// ...and check them to get a package
+	pkg, err = i.config.Check(path, i.fileSet, files, nil)
+	if err != nil {
+		return nil, err
+	}
+	i.packages[ident] = &PackageInfo{
+		BuildInfo: buildPkg,
+		ParseInfo: pkg,
+		Files: files,
+	}
+	log.V(1).Info("imported package from source")
+
+	return pkg, nil
+}
+
+func (i *KindaFastImporter) parsePackageFiles(buildPkg *build.Package) ([]*ast.File, error) {
+	// TODO(directxman12): cgo files, test files, xtest files
+	files := make([]*ast.File, len(buildPkg.GoFiles))
+	errors := make([]error, len(buildPkg.GoFiles))
+
+	var wg sync.WaitGroup
+	for ind, filename := range buildPkg.GoFiles {
+		wg.Add(1)
+		go func(ind int, filepath string) {
+			defer wg.Done()
+			files[ind], errors[ind] = parser.ParseFile(i.fileSet, filepath, nil, 0) // ok to access fset concurrently
```

Comment:
further down main.go we have logic to re-parse the file to update the AST.  That is happening outside of this framework.  That doesn't seem right?

Reply:
we're just parsing, not typechecking, so it's fine.  Most of this logic is for import traversal and whatnot in the typechecking phase -- just parsing can be done mostly in a vacuum.

Or are you suggesting to change it for organizational purposes.